### PR TITLE
Add read only builder

### PR DESCRIPTION
**Pull request by juanbono**
What Changed?
This PR adds a new constructor for what used to be the Clients struct.

Now it is separated into 2 structs:
- without writers, wallet and private key
- with writers, wallet and private key

In this way there is a function for building only the reader part of the clients.

### CODE REVIEW

File: chainio/clients/avsregistry/builder.go [hunk 1]
```diff
@@ -7,6 +7,61 @@ import (
 	"github.com/Layr-Labs/eigensdk-go/logging"
 )
 
+// Build an AVS registry client with the given configuration,
+// HTTP and WS clients, and logger, but without a private key.
+//
+// This is useful for read-only operations, such as fetching metrics.
+func BuildClientsForEcMetrics(
```
Comment (samlaf) [inline chainio/clients/avsregistry/builder.go]: This could be used for many other things than the economic collector metrics. Rename to `BuildReadClients`

File: chainio/clients/elcontracts/builder.go [hunk 2]
```diff
@@ -7,6 +7,36 @@ import (
 	"github.com/Layr-Labs/eigensdk-go/metrics"
 )
 
+// Creates a ClientsBase struct (for read operations) with the given:
+// configuration, HTTP client, logger and metrics.
+func BuildClientsForEcMetrics(
```
Comment (samlaf) [inline chainio/clients/elcontracts/builder.go]:
```suggestion
func BuildReadClients(
```

File: chainio/clients/builder.go [hunk 3]
```diff
@@ -51,11 +51,100 @@ type Clients struct {
...
 }
 
-func BuildAll(
+// This is a struct that holds all the params that are needed to interact
+// with the AVS and EL contracts without the Wallet/Private key.
+// This is useful for read-only operations, such as fetching metrics.
+type ClientsBase struct {
+	AvsRegistryChainReader      *avsregistry.ChainReader
+	AvsRegistryChainSubscriber  *avsregistry.ChainSubscriber
+	AvsRegistryChainWriter      *avsregistry.ChainWriter
```
Comment (samlaf) [inline chainio/clients/builder.go]: this should prob not be here?

File: chainio/clients/builder.go [hunk 4]
```diff
@@ -51,11 +51,100 @@ type Clients struct {
...
+	}
+	return &base, nil
+}
+
+func BuildAll( // BUILDS A ClientsWithPkKey
 	config BuildAllConfig,
 	ecdsaPrivateKey *ecdsa.PrivateKey,
 	logger logging.Logger,
-) (*Clients, error) {
+) (*ClientsWithPkKey, error) {
```
Comment (samlaf) [inline chainio/clients/builder.go]: Is the `Clients` struct still used anywhere? Feels to me like ClientsWithPkKey is meant to replace it? Why not just remove the old Clients struct and replace it with the `ClientWithPkKey` struct?

File: chainio/clients/builder.go [hunk 5]
```diff
@@ -51,11 +51,100 @@ type Clients struct {
...
+	EthWsClient                 eth.WsBackend
+	AvsRegistryContractBindings *avsregistry.ContractBindings
+	EigenlayerContractBindings  *elcontracts.ContractBindings
+	Metrics                     *metrics.EigenMetrics // exposes main avs node spec metrics that need to be incremented by avs code and used to start the metrics server
+	PrometheusRegistry          *prometheus.Registry
+}
+
+// This is a struct that extends ClientsBase with Wallet and TxManager.
+// This is useful for the full set of operations (reads and writes), such as sending transactions.
+type ClientsWithPkKey struct {
```
Comment (samlaf) [inline chainio/clients/builder.go]: Why is this added as a new struct? Is the old `Clients` still used? If not, let's get rid of Clients, and just rename this new struct `Clients`.

File: chainio/clients/builder.go [hunk 6]
```diff
@@ -51,11 +51,100 @@ type Clients struct {
 	PrometheusRegistry          *prometheus.Registry  // Used if avs teams need to register avs-specific metrics
 }
 
-func BuildAll(
+// This is a struct that holds all the params that are needed to interact
+// with the AVS and EL contracts without the Wallet/Private key.
+// This is useful for read-only operations, such as fetching metrics.
+type ClientsBase struct {
```
Comment (samlaf) [inline chainio/clients/builder.go]:
```suggestion
type ReadClients struct {
```

**Review by samlaf — CHANGES_REQUESTED** [review]
Good start. Just need to rename some of the structs.

File: chainio/clients/builder.go (hunk 5 above)
Comment (juanbono) [reply chainio/clients/builder.go]: Done. I renamed `ClientsWithPkKey` to `Clients` to preserve the name.

File: chainio/clients/builder.go (hunk 4 above)
Comment (juanbono) [reply chainio/clients/builder.go]: Done. I renamed `ClientsWithPkKey` to `Clients` to preserve the name and removed the old `Clients` struct definition.

File: chainio/clients/builder.go (hunk 3 above)
Comment (juanbono) [reply chainio/clients/builder.go]: Done

File: chainio/clients/elcontracts/builder.go (hunk 2 above)
Comment (juanbono) [reply chainio/clients/elcontracts/builder.go]: Done.

File: chainio/clients/avsregistry/builder.go (hunk 1 above)
Comment (juanbono) [reply chainio/clients/avsregistry/builder.go]: Done. I renamed it to `BuildReadClients`

**Comment by juanbono** [comment]
I addressed all the comments and added some more docs. cc @samlaf

File: chainio/clients/builder.go [hunk 7]
```diff
@@ -120,22 +185,25 @@ func BuildAll(
 		return nil, utils.WrapError("Failed to create EL Reader and Writer", err)
 	}
 
-	return &Clients{
+	base := ReadClients{
```
Comment (samlaf) [inline chainio/clients/builder.go]:
```suggestion
	readClients := ReadClients{
```

File: chainio/clients/builder.go [hunk 8]
```diff
@@ -29,28 +29,93 @@ type BuildAllConfig struct {
 	PromMetricsIpPortAddress   string
 }
 
-// Clients is a struct that holds all the clients that are needed to interact with the AVS and EL contracts.
-// TODO: this is confusing right now because clients are not instrumented clients, but
-// we return metrics and prometheus reg, so user has to build instrumented clients at the call
-// site if they need them. We should probably separate into two separate constructors, one
-// for non-instrumented clients that doesn't return metrics/reg, and another instrumented-constructor
-// that returns instrumented clients and the metrics/reg.
```
Comment (samlaf) [inline chainio/clients/builder.go:37]: please leave this, we still haven't addressed it afaik.

Comment (juanbono) [reply chainio/clients/builder.go:37]: done

File: chainio/clients/builder.go (hunk 7 above)
Comment (juanbono) [reply chainio/clients/builder.go]: done

**Review by samlaf — DISMISSED** [review]
LGTM!

File: chainio/clients/avsregistry/builder.go [hunk 9]
```diff
@@ -7,6 +7,48 @@ import (
 	"github.com/Layr-Labs/eigensdk-go/logging"
 )
 
+// Build an AVS registry client with the given configuration,
+// HTTP and WS clients, and logger, but without a private key.
+//
+// This is useful for read-only operations, such as fetching metrics.
```
Comment (shrimalmadhur) [inline chainio/clients/avsregistry/builder.go]: what does it mean by fetching metrics here?

Comment (juanbono) [reply chainio/clients/avsregistry/builder.go]: Yes, I think can be confusing. I'll remove that bit.

Comment (juanbono) [reply chainio/clients/avsregistry/builder.go]: Done

**Review by shrimalmadhur — APPROVED** [review]
